src/zarr.v2.cpp

```diff
@@ -0,0 +1,268 @@
+            tile_dims.rows,   // y
+            tile_dims.cols,   // x
+          } },
+        { "dtype", common::sample_type_to_dtype(pixel_type_) },
+        { "fill_value", 0 },
+        { "order", "C" },
+        { "filters", nullptr },
+        { "dimension_separator", "/" },
+    };
+
+    if (blosc_compression_params_.has_value()) {
+        zarray_attrs["compressor"] = blosc_compression_params_.value();
+    } else {
+        zarray_attrs["compressor"] = nullptr;
+    }
+
+    std::string zarray_path =
+      (dataset_root_ / std::to_string(level) / ".zarray").string();
+    common::write_string(zarray_path, zarray_attrs.dump());
+}
+
+void
+zarr::ZarrV2::write_external_metadata_() const
+{
+    namespace fs = std::filesystem;
+    using json = nlohmann::json;
+
+    std::string zattrs_path = (dataset_root_ / "0" / ".zattrs").string();
+    common::write_string(zattrs_path, external_metadata_json_);
+}
+
+void
+zarr::ZarrV2::write_base_metadata_() const
+{
+    namespace fs = std::filesystem;
+    using json = nlohmann::json;
+
+    const json zgroup = { { "zarr_format", 2 } };
+    std::string zgroup_path = (dataset_root_ / ".zgroup").string();
+    common::write_string(zgroup_path, zgroup.dump());
+}
+
+void
+zarr::ZarrV2::write_group_metadata_() const
+{
+    namespace fs = std::filesystem;
+    using json = nlohmann::json;
+
+    json zgroup_attrs;
+    zgroup_attrs["multiscales"] = json::array({ json::object() });
+    zgroup_attrs["multiscales"][0]["version"] = "0.4";
+    zgroup_attrs["multiscales"][0]["axes"] = {
+        {
+          { "name", "t" },
+          { "type", "time" },
+        },
+        {
+          { "name", "c" },
+          { "type", "channel" },
+        },
+        {
+          { "name", "y" },
+          { "type", "space" },
+          { "unit", "micrometer" },
+        },
+        {
+          { "name", "x" },
+          { "type", "space" },
+          { "unit", "micrometer" },
+        },
+    };
+
+    // spatial multiscale metadata
+    if (writers_.empty()) {
+        zgroup_attrs["multiscales"][0]["datasets"] = {
+            {
+              { "path", "0" },
+              { "coordinateTransformations",
+                {
+                  {
+                    { "type", "scale" },
+                    { "scale", { 1, 1, pixel_scale_um_.y, pixel_scale_um_.x } },
+                  },
+                } },
+            },
+        };
+    } else {
+        for (auto i = 0; i < writers_.size(); ++i) {
+            zgroup_attrs["multiscales"][0]["datasets"].push_back({
+              { "path", std::to_string(i) },
+              { "coordinateTransformations",
+                {
+                  {
+                    { "type", "scale" },
+                    {
+                      "scale",
+                      {
+                        std::pow(2, i), // t
+                                        // TODO (aliddell): c?
+                        1,              // z
```

Comment:
We don't include `z` right now, which is [optional in the multiscales ome-ngff spec](https://ngff.openmicroscopy.org/latest/#multiscale-md)). So the 1 actually corresponds to the channel axis (to accommodate multi-channel cameras).

I guess this was probably inherited from moving code, but seems like a good time to delete it.